src/handlers/records-write.ts

```diff
@@ -92,46 +87,42 @@ export class RecordsWriteHandler implements MethodHandler {
       };
     }
 
-    const isLatestBaseState = true;
-    const indexes = await recordsWrite.constructRecordsWriteIndexes(isLatestBaseState);
-
-    // if data is below a certain threshold, we embed the data directly into the message for storage in MessageStore.
-    let messageWithOptionalEncodedData: RecordsWriteMessageWithOptionalEncodedData = message;
-
-    // try to store data, unless options explicitly say to skip storage
-    if (options === undefined || !options.skipDataStorage) {
-      if (dataStream === undefined && newestExistingMessage?.descriptor.method === DwnMethodName.Delete) {
-        return messageReplyFromError(new DwnError(DwnErrorCode.RecordsWriteMissingDataStream, 'No data stream was provided with the previous message being a delete'), 400);
+    try {
+      // we set latestBaseState initially to false, due to the fact that we allow `RecordsWrite` without a data stream in some cases.
+      // depending on rules within `processMessageWithDataStream` or `processMessageWithoutDataStream` it will be set to true to allow for query/reads
+      let isLatestBaseState = false;
+      let messageWithOptionalEncodedData = message as RecordsWriteMessageWithOptionalEncodedData;
+
+      if (dataStream !== undefined) {
+        ({ isLatestBaseState, messageWithOptionalEncodedData } = await this.processMessageWithDataStream(tenant, message, dataStream));
```

Comment:
Optional: I think the `processMessageWithDataStream()` and `processMessageWithoutDataStream()` are vaguely named with prefix "process" is because it attempt to handle both the size greater 30K and under cases. I couldn't come up with better names myself! But if we were to split the logic into separate function based on that condition, the function name could be way clear but I can see why one might think it's refactoring not worth taking. So this is optional.

Reply:
Yeah I was considering going the other direction wrt encoded data vs store data paths... but ultimately chose this path.

Potentially refactoring this down the road could be more clear depending on what additional rules/features touch these paths.